kong/conf_loader/init.lua

```diff
@@ -547,8 +549,114 @@ local function load(path, custom_conf, opts)
+    local allow_bundled_filters = false
+    local allow_user_filters = false
+    local disable_all_filters = false
+    for _, filter in ipairs(conf.wasm_filters) do
+      if filter == "bundled" then
+        allow_bundled_filters = true
+
+      elseif filter == "user" then
+        allow_user_filters = true
+
+      elseif filter == "off" then
+        disable_all_filters = true
+
+      else
+        allowed_filter_names[filter] = true
+      end
+    end
+
+    if disable_all_filters then
+      allowed_filter_names = {}
+      allow_bundled_filters = false
+      allow_user_filters = false
+    end
+
+    ---@type kong.configuration.wasm_filter[]
+    local active_filters = {}
+    ---@type table<string, boolean>
+    local active_filter_names = {}
+
+    local bundled_filter_path = conf_constants.WASM_BUNDLED_FILTERS_PATH
+    if not pl_path.isdir(bundled_filter_path) then
+      local alt_path
+
+      local nginx_bin = nginx_signals.find_nginx_bin(conf)
+      if nginx_bin then
+        alt_path = pl_path.dirname(nginx_bin) .. "/../../../kong/include/filters"
+      end
+
+      if alt_path and pl_path.isdir(alt_path) then
+        log.debug("loading bundled proxy-wasm filters from alt path: %s",
+                  alt_path)
+        bundled_filter_path = alt_path
+
+      else
+        log.warn("Bundled proxy-wasm filters path (%s) does not exist " ..
+                 "or is not a directory. Bundled filters may not be " ..
+                 "available", bundled_filter_path)
+      end
+    end
+
+    local bundled_filters = get_wasm_filters(bundled_filter_path)
+    for _, filter in ipairs(bundled_filters) do
+      if allow_bundled_filters or allowed_filter_names[filter.name] then
+        insert(active_filters, filter)
+        active_filter_names[filter.name] = true
+      end
+    end
+
+    ---@type string[]
+    local conflicts = {}
+
+    local user_filters = get_wasm_filters(conf.wasm_filters_path)
+    for _, filter in ipairs(user_filters) do
+      if allow_user_filters or allowed_filter_names[filter.name] then
+        if active_filter_names[filter.name] then
+          insert(conflicts, filter.name)
+
+        else
+          insert(active_filters, filter)
+          active_filter_names[filter.name] = true
+        end
+      end
+    end
+
+    if #conflicts > 0 then
+      local e = "One or more wasm filters in %q conflicts with a bundled " ..
+                "filter name:" ..
+                "\n\n" ..
+                "%s" ..
+                "\n\n" ..
+                "To remedy this, you may need to remove 'bundled' from " ..
+                "`wasm_filters` and add enable any chosen bundled filters " ..
+                "one-by-one. Alternatively, you may rename one of the " ..
+                "conflicting filters in your `wasm_filters_path` directory " ..
+                "(this is necessary if you want to use both the bundled " ..
+                "filter and the user-supplied filter)."
```

Comment:
This is quite clunky UX, given that users would most likely not override wasm filters by accident — if such a conflict happens, they would probably be doing it on purpose with an updated version or the like. Making them reconfigure the _other_ filters because of that feels odd.

Given that `wasm_filters_path` needs to be explicitly set, I would expect user entries to take precedence over bundled entries. We could simply log an info/notice line reporting that "bundled filter x is being overriden by the one in /this/local/path/x.wasm".

Reply:
Felt it best to start out hyper-defensive, but I'd be okay with just logging and moving on.

Reply:
Revisiting this: if I have bundled filters `A` and `B` and user filters `A` and `B`, there's no way to specify that I want to use bundled-`A` and user-`B` without deleting files (in either the original implemented version or my suggestion above).

In a scenario such as a custom-built container image with overriden filters, let's say someone wants to revert to bundled-`A` via configuration only without editing the image. It's a corner case but I don't think it's super far-fetched — let's say I'm trying out a custom build of a filter and I want to easily switch back and forth, but I can't disable the other user filters.

Would it be problematic to support _full paths to filters_ in `wasm_filters`, which would then take precedence over `bundled` or `user`? This would allow for:

* selecting explicitly which version of a filter we want to run
* having user wasm filters in different folders in case one needs to get them mounted via different volumes, etc.
* not having to set two variables to enable a user filter

I'm thinking basically checking whether the entry starts with `/`, and then treating that as an absolute path.

```
wasm_filters=bundled,user,/custom_filters/myfilter.wasm
```

With this, `wasm_filters_path` would essentially become a convenience in case you have multiple filters, but still probably the preferred way to use it, since it's easier to tell users "mount /my_filters/ in your container and then set `wasm_filters_path=/my_filters`.

One can't easily pull off something like this with Lua plugins because of Lua plugins reasons (Lua paths, rock dependencies, etc), but we don't need to constrain ourselves to the limitations of the `plugins` directive.

Reply:
I think if we were to support full paths in `wasm_filters`, we should just ditch `wasm_filters_path` (and the `user` keyword), and make user filters entirely path-based:

### before
```
wasm_filters = bundled,user
wasm_filters_path = /path/to/my/filters
```

### after

```
wasm_filters = bundled,/path/to/my/filters,/path/to/other/filter.wasm
wasm_filters_path = <unset>
```

Honestly I kinda like the simplicity of this. I could see past-us choosing this if we had bundled filters from the get-go.

---

This doesn't directly solve the filter name collision issue. For that, we need to decide some resolution order and decide whether collisions are an error or just a warning... OR we go the yet-another-DSL route:

I like the idea of renaming/aliasing filters in-place with `<name>=<path>`:

```
# where there is already a bundled filter named `dupe_filter`
wasm_filters = bundled,renamed_filter=/path/to/dupe_filter.wasm
```

If you just want to disable the bundled filter without renaming your own, maybe we give you a means of disabling individual filters from the bundled namespace with `!bundled:<name>`:

```
wasm_filters = bundled,!bundled:dupe_filter,/path/to/dupe_filter.wasm
```

Reply:
As for collisions, I think it's fair to assume that any explicitly given entry overrides the version in `bundled` (as in "more specific overrides more general").

I like the idea of `wasm_filters` accepting both files and directories. I slightly dislike always having to enter `bundled,/my/path` when adding a path.

I know that's sort of how the plugins directive already works, but anything that has "it's own syntax" adds a bit of friction ("was it comma separated? does the order matter?" etc etc... the open questions accumulate; the path of the mini-DSL is the ultimate in that). In that regard, having one directive that's just a folder is always easier to remember and to get right as an end-user, so I'm fond of keeping `wasm_filters_path` around, even if it becomes just a shorthand for adding that path to the `wasm_filters` list.

Reply:
> As for collisions, I think it's fair to assume that any explicitly given entry overrides the version in bundled (as in "more specific overrides more general").

Additionally, you can give the user a warning, like `[warn] ... The user plugin <XXX> has overridden the bundled one`, and allow the execution to proceed.

> If you just want to disable the bundled filter without renaming your own, maybe we give you a means of disabling individual filters from the bundled namespace with !bundled:<name>:
>```
>wasm_filters = bundled,!bundled:dupe_filter,/path/to/dupe_filter.wasm
>```

This seems to add a bit of complexity to the use and implementation of  the`wasm_filters` option.